File: 2022/java/Day04.java
/**
 * Script to find overlapping ranges.
 */
import java.util.ArrayList;
import java.util.List;

import utils.Day04Range;
import utils.Day;

public class Day04 extends Day {

    private List<Day04Range[]> parseData() {
        String line;
        List<Day04Range[]> data;
        String[] rangeStrings;
        Day04Range[] ranges;
        Day04Range range1;
        Day04Range range2;

        // Parse each line of input data into array of Day04Range, and store inside List
        data = new ArrayList<>();
        while (this.scanner.hasNextLine()) {
            line = this.scanner.nextLine();
            rangeStrings = line.split(",");
            range1 = new Day04Range(rangeStrings[0].split("-"));
            range2 = new Day04Range(rangeStrings[1].split("-"));
            ranges = new Day04Range[]{range1, range2};
            data.add(ranges);
        }

        return data;
    }

    private int solveOne(List<Day04Range[]> data) {
        int subsetCount;
        Day04Range range1;
        Day04Range range2;

        // Find total count of complete overlaps, where one range is a subset of another range
        subsetCount = 0;
        for (Day04Range[] ranges : data) {
            range1 = ranges[0];
            range2 = ranges[1];
            // Check if range1 is a subset of range2
            if (range1.getStart() >= range2.getStart() && range1.getEnd() <= range2.getEnd()) {
                subsetCount++;
            // Check if range 2 is a subset of range1
            } else if (range2.getStart() >= range1.getStart() && range2.getEnd() <= range1.getEnd()) {
                subsetCount++;
            }
        }

        return subsetCount;
    }

    private int solveTwo(List<Day04Range[]> data) {
        int overlapCount;
        Day04Range range1;
        Day04Range range2;

        // Find total count of partial overlaps, where one range overlaps another range
        overlapCount = 0;
        for (Day04Range[] ranges : data) {
            range1 = ranges[0];
            range2 = ranges[1];
            // Check if ranges overlap
            if (range1.getStart() <= range2.getEnd() && range1.getEnd() >= range2.getStart()) {
                overlapCount++;
            }
        }

        return overlapCount;
    }

    public static void main(String[] args) {
        Day04 day04;
        List<Day04Range[]> data;
        int subsetCount;
        int overlapCount;

        day04 = new Day04();
        data = day04.parseData();

        subsetCount = day04.solveOne(data);
        System.out.println("part one solution: " + subsetCount);

        overlapCount = day04.solveTwo(data);
        System.out.println("part two solution: " + overlapCount);
    }
}

File: 2022/python/day04.py
"""
Script to find overlapping ranges.
"""
import os

from day import Day


class Range:
    def __init__(self, range: list[str, str]):
        self.start = int(range[0])
        self.end = int(range[1])


class Day04(Day):
    def __init__(self):
        super().__init__()
        self.data = self.read_data()

    def read_data(self) -> list[tuple[Range, Range]]:
        # Read input data from text file
        data = []
        with open(os.path.join(self._directory_path, self._data_filename), "r") as file:
            for line in file:
                # Parse each line into a tuple of Ranges
                range_1, range_2 = line.strip().split(",")
                range_1 = range_1.split("-")
                range_2 = range_2.split("-")
                data.append((Range(range_1), Range(range_2)))

        return data

    def solve_one(self) -> int:
        # Find total count of complete overlaps, where one range is a subset of another range
        subset_count = 0
        for range_1, range_2 in self.data:
            # Check if range_1 is a subset of range_2
            if range_1.start >= range_2.start and range_1.end <= range_2.end:
                subset_count += 1
            # Check if range_2 is a subset of range_1
            elif range_2.start >= range_1.start and range_2.end <= range_1.end:
                subset_count += 1

        return subset_count

    def solve_two(self) -> int:
        # Find total count of partial overlaps, where one range overlaps another range
        return sum(
            # Check if ranges overlap
            range_1.start <= range_2.end and range_1.end >= range_2.start
            for range_1, range_2 in self.data
        )


def main():
    day04 = Day04()

    subset_count = day04.solve_one()
    print(f"part one solution: {subset_count}")

    overlap_count = day04.solve_two()
    print(f"part two solution: {overlap_count}")


if __name__ == "__main__":
    main()
